### ISSUE TITLE
Finetuning while keeping TabICL fixed

### ISSUE THREAD

**Issue by albertcthomas** [soda-inria/tabicl#17]
I wanted to finetune end to end a preprocessing step followed by TabICL while keeping TabICL fixed. The issue I'm having is that to freeze TabICL I am putting it in eval mode as it is usually what one would do. But doing so makes `forward` calls `_inference_forward` which itself uses a `with torch.no_grad()` for the row and col embedders thus discarding the gradients.

**Comment by albertcthomas** [comment]
I'm doing the same as what is done in the code [here](https://github.com/soda-inria/tabicl/blob/main/src/tabicl/train/run.py#L196) to freeze the model but this means I am losing the gradients when doing the forward then.

**Comment by JingangQu** [comment]
Hi @albertcthomas,

To be honest, we haven't thoroughly explored the fine-tuning capabilities of TabICL yet. The current codebase is also not well-suited for fine-tuning at the moment.

That said, fine-tuning could indeed be quite helpful for improving performance. We'll definitely take it into consideration, and I'll get back to you soon.

**Comment by JingangQu** [comment]
In the last pretraining stage of TabICL on datasets with 40K - 60K rows, in order to reduce memory consumption, we will freeze the transformer encoders for column-wise embedding and row-wise interaction. The code you mentioned is for this purpose.

**Comment by albertcthomas** [comment]
Ok got it thanks. You froze the transformer encoders for column wise embedding and row-wise interaction and were not backpropagating through them so having no gradient (eval mode) was OK for you?

**Comment by albertcthomas** [comment]
Yes. Except that if you put your transformer encoders in eval mode in your code you will lose the gradients because eval mode will call `_inference_forward` which will use `with torch.no_grad` [here](https://github.com/soda-inria/tabicl/blob/c4cd0d02feef7a8440bd7f656aa8f356a6ce0982/src/tabicl/model/inference.py#L370)

**Comment by albertcthomas** [comment]
Ok thank you very much for confirming this.
